department-of-veterans-affairs/caseflow
# Allow for VSO Users to Enter Representative Email if Their Own Email is Missing

### Description
Allow for the POA/Representative email field to be filled in by VSO users whenever converting a hearing/hearing request to virtual if no email is on file for the current user.

This PR also introduces a [FeatureToggle](https://github.com/department-of-veterans-affairs/caseflow/wiki/Adding-a-Feature-Flag-with-FeatureToggle) which restricts access to the hearing conversion links on the Case Details page for VSO users. The FeatureToggle is named `vso_virtual_opt_in`.

### Acceptance Criteria
- [ ] When accessing the hearing request conversion form and hearing conversion form as a VSO user, the POA email field is read only if the current user's email is present in the database.
- [ ] If the current user's email is missing in the database, and is inaccessible, the POA email field on both of the aforementioned forms is enabled to allow for the VSO user to enter their email.
- [ ] Unless the FeatureToggle `:vso_virtual_opt_in` is enabled for a VSO user, they will not be able to view the hearings section on the Case Details page (as was the case prior to this feature work).
- [ ] These changes do not alter the forms' behaviors for hearings coordinators.

### Testing Plan
1. Log into Caseflow as BILLIE_VSO. This user should have an email.
2. Grab an appeal without a scheduled hearing and a hearing that is not virtual:
```ruby
a = Appeal.all.filter { |a| a.docket_type == "hearing" && a.hearings.empty?}.first

h = Hearing.where(disposition: nil).reject{|h| h.virtual? || h.scheduled_for_past?}.first
```
3. Go to `/queue/appeals/<a.uuid>`. You should NOT see the hearings section on the page.

4. Enable the `:vso_virtual_opt_in` FeatureToggle:
```ruby
FeatureToggle.enable!(:vso_virtual_opt_in)
```

5. Refresh the page. Click on "Convert to virtual" in the hearings section. The "POA/Representative Email" field should be read-only.

6. Go to `/hearings/<h.uuid>/details`. The "POA/Representative Email" field should be read-only here as well.

7. Log into Caseflow as MANDY_VSO. This user should be missing an email.

8. Repeat steps 5 and 6. On both forms you should be able to enter text into the "POA/Representative Email" field.

8a. As MANDY_VSO, enter an invalid email (ex: `not-valid@email`) into the representative email field on both forms and click out of the text box. Ensure the follow message appears:

<img width="388" alt="email-validation-error" src="https://user-images.githubusercontent.com/99351305/176324372-df62e6fb-d343-470a-8e7d-cafd9b74135a.PNG">

9. Log into Caseflow as BVASYELLOW. This user is a hearings coordinator.

10. Disable the `:vso_virtual_opt_in` FeatureToggle. We want to make sure that even with this FT disabled that hearings coordinators can still view the hearings section on the Case Details page normally.
```ruby
FeatureToggle.disable!(:vso_virtual_opt_in)
```
11. Repeat steps 5 and 6. Make sure that "POA/Representative Email" field is not read-only on both forms as hearings coordinators should not be affected by this update. To repeat step 3, use the Task Action dropdown and select "Convert to Virtual".


### User Facing Changes

**User With Email**
 Queue|Hearings
 ---|---
![localhost_3000_queue_appeals_8b8a0c15-4f70-4259-8bee-a6390286a3d1_tasks_1188_change_hearing_request_type_to_virtual](https://user-images.githubusercontent.com/99351305/176505078-db0a391f-baf1-4853-ab98-7982ad721df9.png)|![localhost_3000_hearings_b8d44921-d672-4b0a-8f1f-27c4c9c20f2a_details](https://user-images.githubusercontent.com/99351305/176505100-4ebc382a-a012-4d0e-87e1-ded72e514cf9.png)



**User Without Email**
 Queue|Hearings
 ---|---
![localhost_3000_queue_appeals_8b8a0c15-4f70-4259-8bee-a6390286a3d1 (1)](https://user-images.githubusercontent.com/99351305/176324135-cd0d4cc2-883b-41e5-9a4a-8198e57a80e4.png)|![localhost_3000_hearings_fb77d965-9007-41b9-b59c-8d9fb2441740_details (3)](https://user-images.githubusercontent.com/99351305/176324144-38521727-2b5b-482d-aeef-488fe5c12bf1.png)


### Code Documentation Updates
